src/drivers/gimbal/gimbal.cpp

```diff
@@ -0,0 +1,542 @@
+
+Gimbal::~Gimbal()
+{
+	/* make sure we are truly inactive */
+	stop();
+
+	::close(_actuator_controls_2_topic);
+	::close(_vehicle_command_sub);
+}
+
+int
+Gimbal::init()
+{
+	int ret = ERROR;
+
+	/* do regular cdev init */
+	if (CDev::init() != OK) {
+		goto out;
+	}
+
+	start();
+	ret = OK;
+
+out:
+	return ret;
+}
+
+int
+Gimbal::probe()
+{
+	return OK;
+}
+
+int
+Gimbal::ioctl(struct file *filp, int cmd, unsigned long arg)
+{
+	switch (cmd) {
+
+	case GIMBALIOCATTCOMPENSATE:
+		_attitude_compensation = (arg != 0);
+		return OK;
+
+	default:
+		/* give it to the superclass */
+		return CDev::ioctl(filp, cmd, arg);
+	}
+}
+
+ssize_t
+Gimbal::read(struct file *filp, char *buffer, size_t buflen)
+{
+	return 0;
+}
+
+void
+Gimbal::start()
+{
+	/* schedule a cycle to start things */
+	work_queue(LPWORK, &_work, (worker_t)&Gimbal::cycle_trampoline, this, 1);
+}
+
+void
+Gimbal::stop()
+{
+	work_cancel(LPWORK, &_work);
+}
+
+void
+Gimbal::cycle_trampoline(void *arg)
+{
+	Gimbal *dev = static_cast<Gimbal *>(arg);
+
+	dev->cycle();
+}
+
+void
+Gimbal::cycle()
+{
+	if (!_initialized) {
+		/* get a subscription handle on the vehicle command topic */
+		_vehicle_command_sub = orb_subscribe(ORB_ID(vehicle_command));
+
+		/* get a publication handle on actuator output topic */
+		struct actuator_controls_s zero_report;
+		memset(&zero_report, 0, sizeof(zero_report));
+		zero_report.timestamp = hrt_absolute_time();
+		_actuator_controls_2_topic = orb_advertise(ORB_ID(actuator_controls_2), &zero_report);
+
+		if (_actuator_controls_2_topic < 0) {
+			warnx("advert err");
+		}
+
+		_initialized = true;
+	}
+
+	bool	updated = false;
+
+	perf_begin(_sample_perf);
+
+	float roll = 0.0f;
+	float pitch = 0.0f;
+	float yaw = 0.0f;
+
+
+	if (_attitude_compensation) {
+		if (_att_sub < 0) {
+			_att_sub = orb_subscribe(ORB_ID(vehicle_attitude));
+		}
+
+		vehicle_attitude_s att;
+
+		orb_copy(ORB_ID(vehicle_attitude), _att_sub, &att);
+
+		roll = -att.roll;
+		pitch = -att.pitch;
+		yaw = att.yaw;
+
+		updated = true;
+	}
+
+	struct vehicle_command_s cmd;
+
+	bool cmd_updated;
+
+	orb_check(_vehicle_command_sub, &cmd_updated);
+
+	if (cmd_updated) {
+
+		orb_copy(ORB_ID(vehicle_command), _vehicle_command_sub, &cmd);
+
+		VEHICLE_MOUNT_MODE mountMode = (VEHICLE_MOUNT_MODE)cmd.param7;
+		debug("cmd: %d, mountMode %d | param1: %8.4f param2: %8.4f", cmd.command, mountMode, (double)cmd.param1, (double)cmd.param2);
+
+		if (cmd.command == VEHICLE_CMD_DO_MOUNT_CONTROL && mountMode == VEHICLE_MOUNT_MODE_MAVLINK_TARGETING) {
+
+			/* Convert to range -1 ... 1, which corresponds to -180deg ... 180deg */
+			roll += 1.0f / M_PI_F * M_DEG_TO_RAD_F * cmd.param1;
+			pitch += 1.0f / M_PI_F * M_DEG_TO_RAD_F * cmd.param2;
+			yaw += 1.0f / M_PI_F * M_DEG_TO_RAD_F * cmd.param3;
+			
+			updated = true;
+
+		}
+
+#if 0
+		// XXX change this to the real quaternion command
+		if (cmd.command == VEHICLE_CMD_DO_MOUNT_CONTROL_QUAT && mountMode == VEHICLE_MOUNT_MODE_MAVLINK_TARGETING) {
```

Comment:
The reason I'm suggesting this is that gimbal lock (no pun intended) is a real issue depending on the mounting location of the Gimbal, and we're trying to do gimbal math in Quaternion space only on the interface level.